packages/runtime-core/src/vnode.ts

```diff
@@ -423,6 +423,12 @@ function createBaseVNode(
   isBlockNode = false,
   needFullChildrenNormalization = false
 ) {
+  // #6978 the children maybe a hoisted array that should be cloned
+  // since it maybe changed during HMR
+  if (__DEV__ && isArray(children)) {
+    children = [...children]
+  }
```

Comment:
Not sure if we should do always this, I wonder if we should change `children` to be cloned: `children = (children as VNode[]).map(deepCloneVNode)` 

I also wonder if the condition is correct, that way we are unnecessarily adding a new array every time, from this example alone we could do something like:
```ts
if (
    __DEV__ &&
    (patchFlag === PatchFlags.HOISTED ||
      patchFlag === PatchFlags.UNKEYED_FRAGMENT) &&
    isArray(children)
  )
```
But not certain if that will fix every case 🤔